NeoML/src/Dnn/DnnReferenceRegister.cpp

```diff
@@ -20,47 +20,73 @@ limitations under the License.
 
 namespace NeoML {
 
-CDnnReferenceRegister::CDnnReferenceRegister( CDnn* _originalDnn ) :
-	learningState( false ),
-	referenceCounter( -1 ),
-	originalDnn( _originalDnn ),
-	originalRandom( new CRandom( _originalDnn->Random() ) )
+CDnnReferenceRegister::CDnnReferenceRegister( IMathEngine& mathEngine, CDnn& dnn ) :
+	CDnnReferenceRegister( new CReferenceDnnInfo( dnn.Random(), nullptr ), mathEngine )
 {
-	NeoAssert( _originalDnn != nullptr );
-	if( originalDnn->referenceDnnRegister.referenceCounter == 0 ) {
-		originalDnn->referenceDnnRegister.learningState = originalDnn->IsLearningEnabled();
-	}
-	originalDnn->referenceDnnRegister.referenceCounter += 1;
+	// Copy dnn using serialization to get the new dnn of necessary life time
+	CMemoryFile file;
+	CArchive archive( &file, CArchive::store );
+	dnn.Serialize( archive );
+	archive.Close();
+	file.SeekToBegin();
+
+	// Allow to Serialize() this time
+	CReferenceDnnInfo* tmp = nullptr;
+	swap( tmp, originalDnn.referenceDnnInfo );
+
+	archive.Open( &file, CArchive::load );
+	originalDnn.Serialize( archive );
+	archive.Close();
+
+	// And revert back the restrictions
+	originalDnn.DisableLearning();
+	swap( tmp, originalDnn.referenceDnnInfo );
+	NeoAssert( originalDnn.referenceDnnInfo != nullptr );
+}
+
+CDnnReferenceRegister::CDnnReferenceRegister( IMathEngine& mathEngine, CArchive& archive, int seed ) :
+	CDnnReferenceRegister( new CReferenceDnnInfo( CRandom( seed ), nullptr ), mathEngine )
+{
+	// Allow to Serialize() this time
+	CReferenceDnnInfo* tmp = nullptr;
+	swap( tmp, originalDnn.referenceDnnInfo );
+
+	NeoAssert( archive.IsLoading() );
+	originalDnn.Serialize( archive );
+	archive.Close();
+
+	// And revert back the restrictions
+	originalDnn.DisableLearning();
+	swap( tmp, originalDnn.referenceDnnInfo );
+	NeoAssert( originalDnn.referenceDnnInfo != nullptr );
+}
+
+CDnnReferenceRegister::CDnnReferenceRegister( CReferenceDnnInfo* referenceDnnInfo, IMathEngine& mathEngine ) :
+	originalDnn( referenceDnnInfo->Random(), mathEngine )
+{
+	// Enable this pointer is to add the restriction to change this dnn.
+	// Used for the original dnn and all reference dnns
+	originalDnn.referenceDnnInfo = referenceDnnInfo;
+	NeoAssertMsg( mathEngine.GetType() == MET_Cpu, "CDnnReferenceRegister: Allowed only for CPU mathEngine" );
 }
 
 CDnnReferenceRegister::~CDnnReferenceRegister()
 {
-	NeoAssertMsg( originalDnn != nullptr || referenceCounter == 0,
-		"CDnnReferenceRegister: delete reference dnns before original dnn" );
-
-	if( originalDnn != nullptr ) {
-		originalDnn->referenceDnnRegister.referenceCounter -= 1;
-	}
-	if( referenceCounter == -1
-		&& originalDnn->referenceDnnRegister.referenceCounter == 0
-		&& originalDnn->referenceDnnRegister.learningState == true )
-	{
-		originalDnn->EnableLearning();
-	}
-
-	if( originalRandom != nullptr ) {
-		delete originalRandom;
-	}
+	NeoAssertMsg( counter == 0, "CDnnReferenceRegister: Cannot be destroyed before any reference dnns" );
```

Comment:
Why not making it smth like CPtr<ReferenceDnnFactory> in every info/reference? User could release it after all necessary dnns created, internal mechanisms live until at least one dnn lives.

Reply:
It should be the `std::shared_ptr` with `atomic` counter .
`CPtr` is not. `CPtr` could store children of `CObject` only.

Yes, its implementation of  `std::shared_ptr`, because it cannot be used in headers.

Reply:
Yes, I meant something like thread-safe CPtr, but written from scratch. Probably not worth it

Reply:
Yes, I also think so